pulse-scheduler: init ethereum contracts, subscriptions, updating prices

tejasbadadare opened the pull request:
## Summary

Initialize the Pulse Scheduler Ethereum contracts with a base layer of functionality. Follows the structure/boilerplate of the existing Pulse contracts.

## Rationale
Add state, functions, and tests for:
- Subscription management
  - Only manager can CRUD subscriptions
- Pushing and consuming price updates
  - An update to a subscription's feeds must have matching timestamps and contain all price IDs for that subscription.
  - Since we plan to make updating permissionless, an update is only allowed if the update criteria is met.
  - Only whitelisted readers can consume the price feeds.
- Viewing active subscriptions (for keepers)

## How has this been tested?

- [ ] Current tests cover my changes
- [x] Added new tests
- [ ] Manually tested the code

## Up next
- Balance management (actually using manager funds)
- Keeper/provider management (payment/registration, etc) 
- Gas strategy configuration
- Gas benchmark tests

jayantk commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/Scheduler.sol:
```diff
@@ -0,0 +1,482 @@
+        // We will validate the trigger conditions and timestamps ourselves
+        // using the returned PriceFeeds.
+        uint64 maxPublishTime = SafeCast.toUint64(block.timestamp);
+        uint64 minPublishTime = maxPublishTime - 10 seconds;
```
i'm not sure 10 seconds is enough. Either way, you probably want this value to be configurable by the admin authority.

tejasbadadare replied:
Yeah true, will set it to 1 minute to allow for retries/flaky RPCs which might result in the tx landing later than expected.

jayantk commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/Scheduler.sol:
```diff
@@ -0,0 +1,482 @@
+                    validPriceId = true;
+                    break;
+                }
+            }
```
fyi instead of doing this loop, you can also check if requestedFeeds[i] is all zero below. (or check 1 field of it which is guaranteed to be nonzero in a valid update). 

That approach also works better in the rare case where someone makes a subscription then tries to read it before it's ever updated.

tejasbadadare replied:
Oh yeah, reduces it from O(n^2) to O(n), nice! One sharp corner here is that if the subscribed feeds change, this function would still let them query old price ids (unless we cleared `_state.priceUpdates[subscriptionId]`). But that's pretty minor, and the stale price would quickly fail age validation.

ali-bahjati replied:
probably you can clear the state for excluded assets.

just thinking in a fancy way another way is that you can always store a "version" for each subscription and then after each update increment it and set all the included price feeds to be of that version. then here you just check the version :D

tejasbadadare replied:
that would be neat :) i guess we only really need to update the price feeds if they were updated, and not if some other attribute was changed (like whitelist). will stick with clearing the old feeds for now

jayantk commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/Scheduler.sol:
```diff
@@ -0,0 +1,482 @@
+    {
+        // TODO: This is gonna be expensive because we're iterating through
+        // all subscriptions, including deactivated ones. But because its a view
+        // function maybe it's not bad? We can optimize this.
```
yeah there may be some maximum size limit on the result (?) may need to paginate this or something like that

I also worry that there's a potential vulnerability here where someone goes and creates a billion subscriptions. If there are too many subscriptions, then maybe the keeper can't find the active ones, and that causes the service to go down.

tejasbadadare replied:
Yeah good shout on a potential max size limit on the result, i'll check on that. If we make addSubscription payable and enforce a minimum balance, we should be able to mitigate people flooding the service with subscriptions hopefully.

jayantk commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/Scheduler.sol:
```diff
@@ -0,0 +1,482 @@
+        emit SubscriptionUpdated(subscriptionId);
+    }
+
+    function deactivateSubscription(
```
should this be a special case of updateSubscription -- then you can also re-activate deactivated subscriptions?

tejasbadadare replied:
Hm, when writing this I had considered deactivating a subscription to be a one-way operation. But it is better UX to allow someone to pause their subscription but maintain the balance and settings, so they can resume later. To make it a special case of updateSubscription, we'd need to move `isActive` from `SubscriptionStatus` to `SubscriptionParams`, which maybe feels a little strange from a data model perspective. I considered adding a `updateSubscriptionIsActive(isActive: bool)`, but I think i prefer just one unified `updateSubscription()`. Will make that change

jayantk commented on target_chains/ethereum/contracts/forge-test/PulseScheduler.t.sol:
```diff
@@ -0,0 +1,896 @@
+        scheduler.updatePriceFeeds(subscriptionId, updateData2, priceIds);
+    }
+
+    function testUpdatePriceFeedsRevertsOnUpdateConditionsNotMet_Deviation()
```
what's with the _ in the name here and in the method above

tejasbadadare replied:
Hahah, they're kind of variants of the update criteria unit tests. I suppose a more standard naming convention would be `testUpdatePriceFeedsRevertsOnDeviationUpdateConditionNotMet`

jayantk reviewed (approved):
looks like a good start to me. Feel free to address my comments in follow up PRs, as this is pretty big and it's generally fine to merge as is.

tejasbadadare commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/SchedulerState.sol:
```diff
@@ -0,0 +1,64 @@
+
+contract SchedulerState {
+    // Maximum number of price feeds per subscription
+    uint8 public constant MAX_PRICE_IDS = 10;
```
Yeah it's an arbitrary number, will raise it to something higher

tejasbadadare commented:
@jayantk addressed your comments in #2585 , @ali-bahjati i'll probably make a new PR to address your feedback since it's already sizable and covers a bunch of disjoint changes

ali-bahjati commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/Scheduler.sol:
```diff
@@ -0,0 +1,482 @@
+            value: pythFee
+        }(updateData, priceIds, minPublishTime, maxPublishTime);
+
+        // Verify all price feeds have the same timestamp
```
a corner case here is when markets are closed for some assets. in these cases using `slot` is better but i doubt we expose it from parse price feeds.

tejasbadadare replied:
oh yeah good callout -- lets use slot for this

ali-bahjati commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/Scheduler.sol:
```diff
@@ -0,0 +1,482 @@
+        status.priceLastUpdatedAt = priceFeeds[0].price.publishTime;
+        status.balanceInWei -= pythFee;
+        status.totalUpdates += 1;
+        status.totalSpent += pythFee;
```
you'll handle publishers rewards and cost in a separate pr right?

ali-bahjati commented on target_chains/ethereum/contracts/forge-test/PulseScheduler.t.sol:
```diff
@@ -0,0 +1,896 @@
+        );
+    }
+
+    function testOptionalWhitelist() public {
```
this kind of tests whielist being disabled. the name of test was confusing.

ali-bahjati replied:
where are you testing that whitelisting works actually? can't find it.

ali-bahjati reviewed (approved):
Nice work! looking forward to the subsequent prs.

tejasbadadare commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/SchedulerState.sol:
```diff
@@ -0,0 +1,64 @@
+        bool updateOnDeviation;
+        uint32 deviationThresholdBps;
+
+        // TODO: add updateOnConfidenceRatio?
```
yeahh sounds good, will leave it out

tejasbadadare commented on target_chains/ethereum/contracts/contracts/pulse/scheduler/SchedulerState.sol:
```diff
@@ -0,0 +1,64 @@
+        // Pyth contract for parsing updates and verifying sigs & timestamps
+        address pyth;
+        // Sub ID -> subscription parameters (which price feeds, when to update, etc)
+        mapping(uint256 => SubscriptionParams) subscriptionParams;
+        // Sub ID -> subscription status (metadata about their sub)
+        mapping(uint256 => SubscriptionStatus) subscriptionStatuses;
+        // Sub ID -> price ID -> latest parsed price update for the subscribed feed
+        mapping(uint256 => mapping(bytes32 => PythStructs.PriceFeed)) priceUpdates;
+        // Sub ID -> manager address
+        mapping(uint256 => address) subscriptionManager;
```
Yeahh that honestly might work better now that theres more state. I was thinking of saving a hop by maintaining multiple mappings. It's not hard to change this, so i'll table this until we build out the rest of the contract
